### PR TITLE
Micro-optimizations

### PR DESCRIPTION
**Pull request by PabloKowalczyk**
This PR adds some micro-optimizations, mostly it swaps `array_key_exists` with much faster `isset`.
Performance gain with large data sets can go up to 15%.

### CODE REVIEW

**Comment by lindyhopchris** [comment]
`array_key_exists` and `isset` aren't equivalent:

```php
$data = ['foo' => null'];
array_key_exists('foo', $data'); // true
isset($data['foo']); // false
```

haven't looked into the failing tests to see if that's why they are failing but you'd need to check that any that you're replacing aren't accepting `null` values at the mo.

File: src/Parser/RelationshipData/ParseRelationshipDataTrait.php [hunk 1]
```diff
@@ -55,7 +55,7 @@ private function parseRelationshipData(
         int $nextLevel,
         string $nextPathPrefix
     ): array {
-        $hasData = array_key_exists(SchemaInterface::RELATIONSHIP_DATA, $description);
```
Comment (neomerx) [inline src/Parser/RelationshipData/ParseRelationshipDataTrait.php:58]: This one is bad. `null` is a valid value for data.

**Comment by neomerx** [comment]
Thanks for the PR, it has some really good catches. However, it will require more efforts. A couple of my thoughts to be on the same page
- I would test replacing getters with direct usage calls separately. I expect insignificant improvement so probably not worth it. Such calls are potentially easily could be optimized by PHP itself.
- optimizations in `asserts` are not needed. They [will be removed in production mode](http://php.net/manual/en/ini.core.php#ini.zend.assertions) anyways.
- optimized order in `if` are good catches :congratulations: 
- isset comments in the PR code

File: src/Parser/Parser.php [hunk 2]
```diff
@@ -482,7 +482,7 @@ public function isNull(): bool
      */
     private function isPathRequested(string $path): bool
     {
-        return array_key_exists($path, $this->paths);
```
Comment (neomerx) [inline src/Parser/Parser.php:485]: This one looks OK.

File: src/Parser/RelationshipData/ParseRelationshipLinksTrait.php [hunk 3]
```diff
@@ -49,8 +49,7 @@ private function parseRelationshipLinks(
         assert(is_bool($addSelfLink) === true || $addSelfLink instanceof LinkInterface);
         assert(is_bool($addRelatedLink) === true || $addRelatedLink instanceof LinkInterface);
 
-        $schemaLinks = array_key_exists(SchemaInterface::RELATIONSHIP_LINKS, $description) === true ?
```
Comment (neomerx) [inline src/Parser/RelationshipData/ParseRelationshipLinksTrait.php:52]: This one looks OK.

**Comment by neomerx** [comment]
Tried to replace `array_key_exists` with `isset` separately (PHP 7.3). Changes are within measurement error in both directions.

**Comment by PabloKowalczyk** [comment]
@neomerx results on my branch: 
screenshot_2019-02-10 comparison from untitled to untitled - blackfire

File: src/Parser/IdentifierAndResource.php [hunk 4]
```diff
@@ -168,35 +168,34 @@ public function getRelationships(): iterable
...
                 $description,
                 $nextLevel,
                 $nextPathPrefix
             );
 
             [$hasLinks, $links] =
-                $this->parseRelationshipLinks($this->getSchema(), $this->getData(), $name, $description);
+                $this->parseRelationshipLinks($this->schema, $this->data, $name, $description);
 
-            $hasMeta = array_key_exists(SchemaInterface::RELATIONSHIP_META, $description);
```
Comment (neomerx) [inline src/Parser/IdentifierAndResource.php:190]: Meta can be any so `array_key_exists` should be used.

**Comment by neomerx** [comment]
The PR has 3 types of changes
1) ditch getters;
2) replace `\array_key_exists` function with `isset` 'language construction' where possible;
3) optimize comparison order in `if` statements.

All 3 combined seem to have a positive effect on performance though it's interesting to understand how they work separately and answer questions
- shall all getters be removed?
- shall all `array_key_exists` calls be replaced?

Also important how performance changes are measured. I use both
- blackfire. Shortly, `cd perf && docker-compose run --rm cli_php blackfire run php -d zend.assertions=-1 /app/sample/sample.php -t=100`
- built-in performance test. Shortly, `cd sample && composer perf-test-php-7-3`
And they give different results :smile: In this case, I tend to trust the second one more. Why? Because of the way blackfire presumably work. It replaces a dozen of PHP calls with own proxies to track how an application works and manages resources. Obviously, even very fast and optimized proxy for `\array_key_exists` is no match for `isset` language construction (which is not a function).

I've already tested separately `3)` and it has a noticeable positive effect and `2)` which seem to have no effect in the built-in test but has with blackfire. Thus my expectation is that `1)` has the biggest impact on performance (though it has to be checked).

I hope it helps to understand my thoughts on this.

File: src/Parser/IdentifierAndResource.php (hunk 4 above)
Comment (PabloKowalczyk) [reply src/Parser/IdentifierAndResource.php:190]: Does `null` meta should use "hasMeta" flow?

**Comment by PabloKowalczyk** [comment]
I'm using "blackfire-way" to measure performance, isn't it preferred way?
It is very interesting that `1)` has the biggest impact on performance - calling a function has some footprint, but i was not aware the impact. Should I try to remove more getters?

**Comment by PabloKowalczyk** [comment]
BTW looks like PHP 7.4 will have special opcode for `\array_key_exists` call - https://github.com/php/php-src/commit/f5044a12ddb6f79d59f72c9c12a6cad36d5f6492#diff-3a8139128d4026ce0cb0c86beba4e6b9

**Comment by neomerx** [comment]
@PabloKowalczyk I've just pushed 3 new branches that start from issue227 with changes
- ditch getters/setters (only for 2 files to see how it affects performance)
- use `isset` instead of `array_key_exists` (all places where applicable)
- changes in `if` statements

here is the summary of multiple `cd sample && composer perf-test-php-7-3`

image

As you can see ditching getters/setters only in 2 files is way more efficient than optimizations for `isset` which are about measuring error. Also it would be irrelevant with PHP 7.4 (thanks for the link btw).

I think changes for `if` should be kept with more ditched getters/setters. Maybe add for `isset`.

Please have a look at changes and let me know what you think.

**Comment by neomerx** [comment]
dropped the rest getters/setters and put final (?) result to `issue227` branch

**Comment by PabloKowalczyk** [comment]
Changes looks good for me. Looks like about 10% performance gain for bigger lists which is great.

**Comment by neomerx** [comment]
released https://github.com/neomerx/json-api/releases/tag/v3.0.1

Thank you for your contributions

**Comment by Ilyes512** [comment]
Not sure if it happend in this mr, but I saw you removed some getters and and setters in the Parser.php file. Thats fine and all, but it is a backwards compatible breaking change. A class that extends the Parser and that uses the setters/getters will break. There is also no alternative as you use private scopes properties (something I am also against, because it makes extending a class impossible. Something should be private only if you really don't want someone to use it).

(Also squashing commits is something I really don't like, because you are destroying the history (if you used good commit messages that is).)

**Comment by neomerx** [comment]
@Ilyes512 Firstly, all the components of the library work with interfaces so there are no dependencies from methods other than defined by interfaces. No interfaces were changed in this commit. Secondly, some compatibility is guaranteed for classes. Such guarantees are covered by a special set of classes called 'Extensions'. If someone wants to rely on a particular implementation feature they better to discuss the inclusion to 'Extensions' tests. No tests were changed in this commit. It might sound tough but I can't assume that my every move can break something I don't even know about. It should be some reasonable boundaries of guarantees.

**Pull request closed without merging**